File: frontend/src/types/physics.ts
// ── Projectile Motion ─────────────────────────────────────────────────────────

export interface ProjectileRequest {
  initial_velocity: number;
  launch_angle: number;
  gravity?: number;
  air_resistance?: number;
}

export interface TrajectoryPoint {
  t: number;
  x: number;
  y: number;
  vx: number;
  vy: number;
}

export interface ProjectileResult {
  flight_time: number;
  max_height: number;
  range: number;
  time_to_peak: number;
  impact_velocity: number;
  trajectory: TrajectoryPoint[];
  velocity_x_series: [number, number][];
  velocity_y_series: [number, number][];
}

export interface ProjectileResponse {
  experiment_id: string;
  parameters: Required<ProjectileRequest>;
  results: ProjectileResult;
}

// ── Newton's Laws ─────────────────────────────────────────────────────────────

export interface NewtonRequest {
  force: number;
  mass: number;
  friction_coefficient: number;
  duration?: number;
}

export interface NewtonResult {
  net_force: number;
  acceleration: number;
  final_velocity: number;
  displacement: number;
  velocity_series: [number, number][];
  acceleration_series: [number, number][];
  position_series: [number, number][];
}

export interface NewtonResponse {
  experiment_id: string;
  parameters: Required<NewtonRequest>;
  results: NewtonResult;
}

// ── Pendulum ──────────────────────────────────────────────────────────────────

export interface PendulumRequest {
  length: number;
  mass: number;
  initial_angle: number;
  gravity?: number;
}

export interface PendulumResult {
  period: number;
  frequency: number;
  angular_frequency: number;
  max_velocity: number;
  max_kinetic_energy: number;
  angle_series: [number, number][];
  angular_velocity_series: [number, number][];
  energy_series: [number, number][];
}

export interface PendulumResponse {
  experiment_id: string;
  parameters: Required<PendulumRequest>;
  results: PendulumResult;
}

File: backend/app/schemas/physics.py
from pydantic import BaseModel, Field


# ── Projectile Motion ──────────────────────────────────────────────────────────

class ProjectileRequest(BaseModel):
    initial_velocity: float = Field(..., ge=1, le=200, description="Launch speed (m/s)")
    launch_angle: float = Field(..., ge=1, le=89, description="Launch angle (degrees)")
    gravity: float = Field(9.81, ge=1.0, le=25.0, description="Gravitational acceleration (m/s²)")
    air_resistance: float = Field(0.0, ge=0.0, le=1.0, description="Air resistance coefficient (0–1)")


class ProjectileResultResponse(BaseModel):
    flight_time: float
    max_height: float
    range: float
    time_to_peak: float
    impact_velocity: float
    trajectory: list[dict]
    velocity_x_series: list[list[float]]
    velocity_y_series: list[list[float]]


class ProjectileResponse(BaseModel):
    experiment_id: str
    parameters: ProjectileRequest
    results: ProjectileResultResponse


# ── Newton's Laws ──────────────────────────────────────────────────────────────

class NewtonRequest(BaseModel):
    force: float = Field(..., ge=0.1, le=1000, description="Applied force (N)")
    mass: float = Field(..., ge=0.1, le=500, description="Object mass (kg)")
    friction_coefficient: float = Field(..., ge=0.0, le=1.0, description="Kinetic friction coefficient")
    duration: float = Field(5.0, ge=1.0, le=30.0, description="Simulation duration (s)")


class NewtonResultResponse(BaseModel):
    net_force: float
    acceleration: float
    final_velocity: float
    displacement: float
    velocity_series: list[list[float]]
    acceleration_series: list[list[float]]
    position_series: list[list[float]]


class NewtonResponse(BaseModel):
    experiment_id: str
    parameters: NewtonRequest
    results: NewtonResultResponse


# ── Pendulum ───────────────────────────────────────────────────────────────────

class PendulumRequest(BaseModel):
    length: float = Field(..., ge=0.1, le=10.0, description="Pendulum string length (m)")
    mass: float = Field(..., ge=0.01, le=10.0, description="Bob mass (kg)")
    initial_angle: float = Field(..., ge=1.0, le=89.0, description="Initial angle from vertical (degrees)")
    gravity: float = Field(9.81, ge=1.0, le=25.0, description="Gravitational acceleration (m/s²)")


class PendulumResultResponse(BaseModel):
    period: float
    frequency: float
    angular_frequency: float
    max_velocity: float
    max_kinetic_energy: float
    angle_series: list[list[float]]
    angular_velocity_series: list[list[float]]
    energy_series: list[list[float]]


class PendulumResponse(BaseModel):
    experiment_id: str
    parameters: PendulumRequest
    results: PendulumResultResponse
